[FSDP] Use unflatten_parameter in _summon_full_parameters

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72467

`_unflatten_params_as_views` does not (and cannot) delete `flat_param`. Using `_unflatten_params_as_views` to implement `_summon_full_parameters` will make the traverse of the full parameter failed -- will get the redundent flat_param.

Differential Revision: [D33989893](https://our.internmc.facebook.com/intern/diff/D33989893/)

fegin committed:
[FSDP] Use unflatten_parameter in _summon_full_parameters

`_unflatten_params_as_views` does not (and cannot) delete `flat_param`. Using `_unflatten_params_as_views` to implement `_summon_full_parameters` will make the traverse of the full parameter failed -- will get the redundent flat_param.

Differential Revision: [D33989893](https://our.internmc.facebook.com/intern/diff/D33989893/)

[ghstack-poisoned]

rohan-varma commented on test/distributed/fsdp/test_fsdp_summon_full_params.py:
```diff
@@ -1,12 +1,13 @@
 # Owner(s): ["oncall: distributed"]
-import sys
 import math
+import sys
```
nit: Can we have formatting changes in separate diffs in the future, as it makes the diff a lot easier to review :) thanks!

rohan-varma commented on test/distributed/fsdp/test_fsdp_summon_full_params.py:
```diff
@@ -97,27 +95,25 @@ def test_summon_full_param_shard_value(self):
             self.assertEqual(raw_model_size, self.get_model_param_count(model))
-            all_shards = next(model.parameters())
+            parameters = list(model.parameters())
+            all_shards = FlatParameter(parameters, requires_grad=False)
```
why do we need to wrap it with `FlatParameter` here?

rohan-varma commented on test/distributed/fsdp/test_fsdp_summon_full_params.py:
```diff
@@ -271,6 +286,7 @@ def test_params_are_unflatenned(self):
             b = flattened_param.detach()
             self.assertTrue(torch.equal(a, b))
 
+
```
nit: extra line

rohan-varma commented on torch/distributed/fsdp/flatten_params_wrapper.py:
```diff
@@ -197,6 +199,10 @@ def shard_metadata(
         )
 
+    @property
+    def param_info(self) -> Tuple[ParamInfo]:
```
Add a small docstring explaining what this is?

rohan-varma commented on torch/distributed/fsdp/flatten_params_wrapper.py:
```diff
@@ -287,14 +298,17 @@ def _init_flatten_params(
         return params, param_infos, shared_param_infos
 
-    def _flatten_params(self) -> None:
+    def _flatten_params(self, external_data: Optional[FlatParameter] = None) -> None:
```
Below doc needs to be updated to reflect `external_data` parameter?

rohan-varma commented on torch/distributed/fsdp/flatten_params_wrapper.py:
```diff
@@ -287,14 +298,17 @@ def _init_flatten_params(
-        ), "Can not flatten params when flat_param is None"
+            getattr(self, "flat_param", None) is not None or external_data is not None
+        ), "Can not flatten params when both flat_param and external_data are None."
+        self.flat_param = (
```
Can this be simplified to

```
if external_data:
    self.flat_param = external_data
```

rohan-varma commented on torch/distributed/fsdp/flatten_params_wrapper.py:
```diff
@@ -339,6 +353,26 @@ def _unflatten_params(self) -> None:
+        Unflatten params. If the current instance is already unflattened, then
+        it will remain unflattened after the context manager exits.
+        """
+        orig_flattened = self.flat_param is not None
```
Would the code be simpler if we added a check and just returned a noop context manager if orig_flattened = False? Such as

```
orig_flattened = ...
if not orig_flattened:
    return suppress()
# Rest of code
```

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -740,16 +743,18 @@ def _summon_full_params(self, recurse: bool = True, writeback: bool = True) -> G
-                self._free_full_params(currently_local_params)
-                self._use_param_local_shard()
-                self.training_state = TrainingState_.IDLE
+            with contextlib.ExitStack() as stack:
```
nit: Add a small comment about what's going on here?

rohan-varma commented on test/distributed/fsdp/test_fsdp_summon_full_params.py:
```diff
@@ -249,7 +263,9 @@ def test_reshard_outside_forward_backward_iteration(self):
         with model._summon_full_params():
             pass
-        self.assertEqual(outer_full_param_size, outer_param._full_param_padded.storage().size())
+        self.assertEqual(
```
Couldn't see if this was already added, but can we add a test that the parameter names are fixed / back to the original module's with this unflattening fix?

fegin replied:
I added a test to check the number of `fsdp_model.parameters()` and the value of the parameters. Testing keys is more tricky as this relies on `state_dict()`  or `named_parameter()` which both do not call `_summon_full_parameters()` in the current code base.

rohan-varma replied:
I see, I filed an issue to track this to make sure that we fix it: https://github.com/pytorch/pytorch/issues/72716

rohan-varma reviewed (commented):
Looks great overall! Just some small comments/questions before stamping

zhaojuanmao commented on torch/distributed/fsdp/flatten_params_wrapper.py:
```diff
@@ -243,6 +249,11 @@ def __init__(self, module: nn.Module, param_list: List[nn.Parameter]):
+        # unflatten_params() context. Using an attribute instead of a
+        # local variable allows us to access the flat_param metadata
+        # even if flat_param is temporarily deleted.
+        self._orig_flat_param = [None]
```
curious why local variable '_orig_flat_param' can not access the flat_param object after self.flat_param is deleted? iiuc, local variable '_orig_flat_param' will still point to the object, 'del self.flat_param' is just reduce the reference counts of the object.

zhaojuanmao replied:
that makes sense!

fegin commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -740,16 +743,18 @@ def _summon_full_params(self, recurse: bool = True, writeback: bool = True) -> G
 
             currently_local_params = self._collect_local_params()
             self._rebuild_full_params()
-            self._fsdp_wrapped_module._unflatten_params_if_needed()
```
It's still used in forward.

zhaojuanmao replied:
I see, keep it then

fegin committed:
Update on "[FSDP] Use unflatten_parameter in _summon_full_parameters"


`_unflatten_params_as_views` does not (and cannot) delete `flat_param`. Using `_unflatten_params_as_views` to implement `_summon_full_parameters` will make the traverse of the full parameter failed -- will get the redundent flat_param.

Differential Revision: [D33989893](https://our.internmc.facebook.com/intern/diff/D33989893/)

[ghstack-poisoned]

rohan-varma commented on test/distributed/fsdp/test_fsdp_summon_full_params.py:
```diff
@@ -271,6 +287,18 @@ def test_params_are_unflatenned(self):
             self.assertTrue(torch.equal(a, b))
 
+    def test_params_count_and_value(self):
+        model = nn.Sequential(
```
you can also import and test `NestedWrappedModel` to test a nested model.

rohan-varma reviewed (approved):
LGTM

Please ensure @zhaojuanmao 's comments are resolved as well :)

zhaojuanmao commented:
lgtm as well, thanks for fixing it!

fegin committed:
Update on "[FSDP] Use unflatten_parameter in _summon_full_parameters"


`_unflatten_params_as_views` does not (and cannot) delete `flat_param`. Using `_unflatten_params_as_views` to implement `_summon_full_parameters` will make the traverse of the full parameter failed -- will get the redundent flat_param.

Differential Revision: [D33989893](https://our.internmc.facebook.com/intern/diff/D33989893/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Use unflatten_parameter in _summon_full_parameters"


`_unflatten_params_as_views` does not (and cannot) delete `flat_param`. Using `_unflatten_params_as_views` to implement `_summon_full_parameters` will make the traverse of the full parameter failed -- will get the redundent flat_param.

Differential Revision: [D33989893](https://our.internmc.facebook.com/intern/diff/D33989893/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Use unflatten_parameter in _summon_full_parameters"


`_unflatten_params_as_views` does not (and cannot) delete `flat_param`. Using `_unflatten_params_as_views` to implement `_summon_full_parameters` will make the traverse of the full parameter failed -- will get the redundent flat_param.

Differential Revision: [D33989893](https://our.internmc.facebook.com/intern/diff/D33989893/)

[ghstack-poisoned]
